### PR TITLE
Fix Out Of Direct Memory Exception and FileInfo/TransferInfo

### PR DESCRIPTION
**Pull request by fredericBregier**
Under certain circumstances, Direct Memory reach a maximum of usage.
This fix tries to limit as possible this impact.
    
- File copy: no more usage of `Guava` where big block where used (512KB) and reduced direct memory (`FileChannel`)
- File reading: now using standard buffer and no more direct memory
- Hash computing: uses a byte array instead of `ByteBuf` from Netty, except when necessary
- KeepAlive algorithm fixed where under certain circumstances it failed while long task delay could lead to TimeOut erroneously
- Free quickier some buffers from Netty
- For `LocalPacket`, makes as possible unique buffer from default allocator (pooled allocation). The `DataPacket`, most often used, is optimized regarding the way of transfer (`ByteBuf` for receiving, byte array for sending).
- `NetworkPacket` was a wrapped buffer of multiple sub-buffers (from `LocalPacket`), turning it to a single Netty `ByteBuf` when reasonable.
- Direct Memory was reduced as much as possible, including Netty configuration itself (_see below_).
- One big and long IT test added to check the correctness of this (no Out Of Direct Memory Error detected)

In Netty, most of the `ByteBuf` are allocated using a Pool Allocator. Depending on usage (IO such as from network, application such as standard allocation), those could be Direct or Heap based. Netty ensures to allocate as needed and handles memory allocation in low consumption, such as Garbage collector is not called too often.
However, if the application uses intensively Direct allocation itself (not through Netty), it could lead to memory issue between Netty and the application.
This fix reduces a lot the Direct allocation out of Netty, increases the usage of Netty Pool Allocator and therefore decrease the memory pressure.

Direct Memory was reduced as much as possible.
**Default configuration is adapted at startup but could be override:**
- `-Dio.netty.noPreferDirect=true` is setup by default, it limits the usage of Direct Buffer except Netty network incoming messages
- `-Dio.netty.maxDirectMemory=0` is setup by default, Netty will use the max JDK Direct memory (recommended).
    
_**Recommended value for `io.netty.maxDirectMemory` is 0**_. Giving too small value (>0) coud cause issues.
- Value of -1 is default behavior of Netty and is also a good option for R66 (Netty uses twice Direct memory maximum from JDK without Cleaner).
- Value of 0 lets Netty allocate within Direct Memory, using JDK max.
- Value of greater than 0 should not be used (could limit too much memory allocation for Netty) and **is really not recommended**.    
  
Note however that Direct memory is still used by Netty itself for incoming messages. Disallowing totally Direct buffer could leads to serious issues on memory usage (GC overacting). Therefore, extra arguments such as `-Dio.netty.noUnsafe=true` **is intensively not recommended**.

As previously, one could also limit usage of Direct Memory by specifying in XML configuration file in `limit` section the value `usenio` to `False`.

**FileInfo/TransferInfo**
In some old revision, an inversion was done on FileInfo and TransferInfo.

FileInfo is the information given by the end user through `-info` and is linked to the file transfer.
This information contains the `FollowId` field (as Json Map) to allow to forward this info in retransfer.

TransferInfo is the internal information, stored by R66 servers and clients, into database about the transfer.
It could be the original size, some special information (such as Digest, UUID) but also the Follow ID.
Its internal representation is a JSON Map saved as a String (so escaped).

This fixes this partial inversion and also improves the HTML rendering by explicitly presenting the `Follow Id`.
This fixes also some cases where the Follow Id was ignored during retransfer (`TransferTask`).

### CODE REVIEW

File: WaarpCommon/src/main/java/org/waarp/common/utility/SystemPropertyUtil.java [hunk 1]
```diff
@@ -94,6 +109,41 @@ public static void refresh() {
...
+        Strings.isNullOrEmpty(get(IO_NETTY_NOPREFERDIRECT))) {
+      try {
+        System.setProperty(IO_NETTY_NOPREFERDIRECT, "true");
+        synchronized (PROPS) {
+          PROPS.clear();
+          PROPS.putAll(newProps);
+        }
+        SysErrLogger.FAKE_LOGGER.syserr(IO_NETTY_NOPREFERDIRECT + ":" +
+                                        io.netty.util.internal.SystemPropertyUtil
+                                            .get(IO_NETTY_NOPREFERDIRECT));
```
Comment (bcarlin) [inline WaarpCommon/src/main/java/org/waarp/common/utility/SystemPropertyUtil.java]: id.

File: WaarpCommon/src/main/java/org/waarp/common/utility/SystemPropertyUtil.java [hunk 2]
```diff
@@ -94,6 +109,41 @@ public static void refresh() {
           PROPS.clear();
           PROPS.putAll(newProps);
         }
+        SysErrLogger.FAKE_LOGGER.syserr(IO_NETTY_ALLOCATOR_TYPE + ":" +
+                                        io.netty.util.internal.SystemPropertyUtil
+                                            .get(IO_NETTY_ALLOCATOR_TYPE));
```
Comment (bcarlin) [inline WaarpCommon/src/main/java/org/waarp/common/utility/SystemPropertyUtil.java]: This is for debugging. It belongs more to the log file than to the error output...

File: WaarpCommon/src/main/java/org/waarp/common/utility/SystemPropertyUtil.java [hunk 3]
```diff
@@ -94,6 +109,41 @@ public static void refresh() {
...
+        Strings.isNullOrEmpty(get(IO_NETTY_MAXDIRECTMEMORY))) {
+      try {
+        System.setProperty(IO_NETTY_MAXDIRECTMEMORY, "0");
+        synchronized (PROPS) {
+          PROPS.clear();
+          PROPS.putAll(newProps);
+        }
+        SysErrLogger.FAKE_LOGGER.syserr(IO_NETTY_MAXDIRECTMEMORY + ":" +
+                                        io.netty.util.internal.SystemPropertyUtil
+                                            .get(IO_NETTY_MAXDIRECTMEMORY));
```
Comment (bcarlin) [inline WaarpCommon/src/main/java/org/waarp/common/utility/SystemPropertyUtil.java]: id.

**Review by bcarlin — CHANGES_REQUESTED** [review]
That's a lot of optimizations.
I did not spot any obvious errors... just a few output that I think should be in the logs (I think they already are in netty initialization logs...): it is always unsettling to have output starting with "ERROR" when the server starts ok.

Other than that, in my test, it fixed the OODM errors I had before and stabilized the RAM usage.
good job !

File: WaarpCommon/src/main/java/org/waarp/common/utility/SystemPropertyUtil.java (hunk 1 above)
Comment (fredericBregier) [reply WaarpCommon/src/main/java/org/waarp/common/utility/SystemPropertyUtil.java]: Agree

**Comment by fredericBregier** [comment]
@bcarlin All done and checked

**Comment by bcarlin** [comment]
I guess this PR can be closed, as it is included in #82 ?

**Pull request merged**